### PR TITLE
new bulk action for checksum reporting

### LINKED ISSUE #1530: back end for checksum report bulk action
(part of uber ticket sul-dlss/argo#1145)

Implementation of ChecksumReportJob (ActiveJob) which will take a list of druids and make one or more ReST calls to preservation_catalog API endpoint (see sul-dlss/preservation_catalog/issues/1202) to get checksum report(s) for said list of druids.

Assume there will be a prescat ReST API that will take one or more druids and return the checksums in csv format.

### PR DESCRIPTION
**Pull request by peetucket**
fixes #1530 

To Do:
~~Fix new ChecksumReportJob test so it doesn't fail:~~

- [x] add expectation to confirm the pres cat API call is made
- [x] add expectation to confirm the CSV file is written
- [x] remove debuging `raise` statement in ChecksumReportJob class (around L35)

~~Get PO Approval~~

### CODE REVIEW

File: app/jobs/checksum_report_job.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require 'faraday'
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: I don't believe you need this here. It should already be required on line 14 of `config/application.rb`.

File: app/jobs/checksum_report_job.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  ##
+  # A job that, given list of pids of objects, runs a checksum report using a presevation_catalog API endpoint and returns a CSV file to the user
+  # @param [Integer] bulk_action_id GlobalID for a BulkAction object
+  # @param [Hash] params additional parameters that an Argo job may need
+  # @option params [Array] :pids required list of pids
+  # @option params [Array] :groups the groups the user belonged to when the started the job. Required for because groups are not persisted with the user.
+  # @option params [Array] :user the user
+  def perform(bulk_action_id, params)
+    super
+    pids = params[:pids]
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: `GenericJob` already has an attr for `pids` that pulls from params, so you can remove this line.

File: app/jobs/checksum_report_job.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    report_filename = generate_report_filename(params[:output_directory])
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: Can you add this param to the yard doc above?

File: app/jobs/checksum_report_job.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    with_bulk_action_log do |log|
+      log.puts("#{Time.current} Starting #{self.class} for BulkAction #{bulk_action_id}")
+      update_druid_count
+      begin
+        raise "#{Time.current} ChecksumReportJob not authorized to view all content}" unless ability.can?(:view_content, ActiveFedora::Base)
+
+        response = conn.get '/objects/checksums', druids: pids, format: 'csv'
+        raise "Call to preservation_catalog_api failed: #{response.status} #{response.reason_phrase}" unless response.success?
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: Depending on how informative prescat's error messages are, you may also want to consider adding `response.body` to the message here.

File: app/jobs/checksum_report_job.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      log.puts("#{Time.current} Finished #{self.class} for BulkAction #{bulk_action_id}")
+    end
+  end
+
+  ##
+  # Generate a filename for the job's csv report output file.
+  # @param  [String] output_dir Where to store the csv file.
+  # @return [String] A filename for the csv file.
+  def generate_report_filename(output_dir)
+    FileUtils.mkdir_p(output_dir) unless File.directory?(output_dir)
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: Did you find that this was needed? I thought the `BulkActionPersister` would do this for us already.

File: app/jobs/checksum_report_job.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        raise "#{Time.current} ChecksumReportJob not authorized to view all content}" unless ability.can?(:view_content, ActiveFedora::Base)
+
+        response = conn.get '/objects/checksums', druids: pids, format: 'csv'
+        raise "Call to preservation_catalog_api failed: #{response.status} #{response.reason_phrase}" unless response.success?
+
+        File.open(report_filename, 'w') { |file| file.write(response.body) }
+        bulk_action.update(druid_count_success: pids.length) # this whole job is run in one call, so it either all succeeds or fails
+      rescue StandardError => e
+        bulk_action.update(druid_count_fail: pids.length)
+        log.puts("#{Time.current} ChecksumReportJob creation failed #{e.class} #{e.message}")
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: Worth tossing this up to honeybadger too?

Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: Yup, but only tossed up the failed call to preservation_catalog API ... other failed jobs don't notify HB currently, but our internal API call failing seems worth a notification

File: app/jobs/checksum_report_job.rb (hunk 5 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: seems to be required -- doesn't work without it

descmetadatadownload does it too: https://github.com/sul-dlss/argo/blob/master/app/jobs/descmetadata_download_job.rb#L99-L102

File: app/jobs/checksum_report_job.rb (hunk 4 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: good idea, done

File: app/jobs/checksum_report_job.rb (hunk 3 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: done

File: app/jobs/checksum_report_job.rb (hunk 2 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: done

File: app/jobs/checksum_report_job.rb (hunk 1 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: done

File: app/jobs/checksum_report_job.rb (hunk 5 above)
Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: but now i wonder if its only there for the tests, since we don't use the BulkActionPersister

Comment (peetucket) [reply app/jobs/checksum_report_job.rb]: in which case, perhaps the test should just call that line in BulkActionPersister specifically so we can remove the directory create line from the job class itself (though it it just duplicative and is not harmful to have)

File: app/jobs/checksum_report_job.rb [hunk 7]
```diff
@@ -11,12 +9,12 @@ class ChecksumReportJob < GenericJob
   # A job that, given list of pids of objects, runs a checksum report using a presevation_catalog API endpoint and returns a CSV file to the user
   # @param [Integer] bulk_action_id GlobalID for a BulkAction object
   # @param [Hash] params additional parameters that an Argo job may need
-  # @option params [Array] :pids required list of pids
-  # @option params [Array] :groups the groups the user belonged to when the started the job. Required for because groups are not persisted with the user.
-  # @option params [Array] :user the user
+  # @option params [Array]  :pids required list of pids
+  # @option params [Array]  :groups the groups the user belonged to when the started the job. Required for because groups are not persisted with the user.
+  # @option params [Array]  :user the user
+  # @option params [String] :the output directory to write the CSV checksum report to
```
Comment (mjgiarlo) [inline app/jobs/checksum_report_job.rb]: Typo: `s/:the/:output_directory the/`

File: spec/jobs/checksum_report_job_spec.rb [hunk 8]
```diff
@@ -6,13 +6,14 @@
   let(:pids) { ['druid:123', 'druid:456'] }
   let(:groups) { [] }
   let(:user) { instance_double(User, to_s: 'jcoyne85') }
-  let(:output_directory) { 'tmp/success' }
-  let(:output_directory_fail) { 'tmp/fail' }
+  let(:output_directory) { 'tmp/checksum_report_job_success' }
+  let(:output_directory_fail) { 'tmp/checksum_report_job_fail' }
+  # different output_directory so our 'fail' test doesn't inadvertantely fail due to the CSV already existing from the 'success' test
```
Comment (mjgiarlo) [inline spec/jobs/checksum_report_job_spec.rb]: Typo: inadvertently

File: app/jobs/checksum_report_job.rb (hunk 5 above)
Comment (mjgiarlo) [reply app/jobs/checksum_report_job.rb]: Could be. It's pretty harmless, so no biggie.

File: spec/jobs/checksum_report_job_spec.rb (hunk 8 above)
Comment (peetucket) [reply spec/jobs/checksum_report_job_spec.rb]: thanks...fixed both, rebased, pushed...

**Comment by ndushay** [comment]
We have approval from Andrew.  w00t!

File: app/views/bulk_actions/_checksum_report_job.html.erb [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<% if bulk_action.present? && bulk_action.status == 'Completed' && File.exists?(File.join(bulk_action.output_directory, Settings.CHECKSUM_REPORT_JOB.CSV_FILENAME)) %>
```
Comment (jcoyne) [inline app/views/bulk_actions/_checksum_report_job.html.erb]: Can this conditional be moved to a helper or controller method?  Prefer to keep code in templates to a minimum.

File: config/settings.yml [hunk 10]
```diff
@@ -12,6 +12,9 @@ BULK_METADATA:
   XML: 'MODS'
   ZIP: 'desc_metadata_download.zip'
 
+CHECKSUM_REPORT_JOB:
```
Comment (jcoyne) [inline config/settings.yml]: Please lowercase these.  The config gem documents the correct usage as lowercase: https://github.com/railsconfig/config#embedded-ruby-erb  We have lowercase variables already in this file.  If we use uppercase the config gem ends up creating methods with the ruby convention for constants, which is confusing.

File: spec/jobs/checksum_report_job_spec.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      it 'calls the presevation_catalog API, writes a CSV file, and records success counts' do
+        response = instance_double(Faraday::Response, success?: true, body: csv_response)
+        allow(my_conn).to receive(:get).and_return(response)
+        subject.perform(bulk_action.id,
+                        output_directory: output_directory,
+                        pids: pids,
+                        groups: groups,
+                        user: user)
+        expect(my_conn).to have_received(:get).with('/objects/checksums', druids: pids, format: 'csv')
+        expect(File.exist?(File.join(output_directory, Settings.CHECKSUM_REPORT_JOB.CSV_FILENAME))).to be_truthy
```
Comment (jcoyne) [inline spec/jobs/checksum_report_job_spec.rb]: This should be `expect(File).to exist(...)`  See https://www.rubydoc.info/gems/rspec-expectations/RSpec/Matchers%3Aexist

File: app/views/bulk_actions/_checksum_report_job.html.erb (hunk 9 above)
Comment (jcoyne) [reply app/views/bulk_actions/_checksum_report_job.html.erb]: I think the safe navigation operator could be used here too (e.g. `bulk_action&.status`) allowing you to drop the first clause

File: spec/jobs/checksum_report_job_spec.rb (hunk 11 above)
Comment (mjgiarlo) [reply spec/jobs/checksum_report_job_spec.rb]: And if you make this change you can remove the new rspec predicate matcher exclusions from `.rubocop_todo.yml` :point_up:

File: config/settings.yml (hunk 10 above)
Comment (ndushay) [reply config/settings.yml]: sure.

**Comment by ndushay** [comment]
@mjgiarlo @jcoyne ready for re-review

File: .rubocop_todo.yml [hunk 12]
```diff
@@ -220,6 +220,14 @@ RSpec/NamedSubject:
 RSpec/NestedGroups:
   Max: 6
 
+# Offense count: 3
+# Cop supports --auto-correct.
+# Configuration parameters: Strict, EnforcedStyle.
+# SupportedStyles: inflected, explicit
+RSpec/PredicateMatcher:
+  Exclude:
+    - 'spec/helpers/bulk_action_helper_spec.rb'
```
Comment (ndushay) [inline .rubocop_todo.yml]: I tried fixing this, but ... see below

File: spec/helpers/bulk_action_helper_spec.rb [hunk 13]
```diff
@@ -32,4 +32,40 @@
...
+    let(:bulk_action) { create(:bulk_action, action_type: 'ChecksumReportJob') }
+
+    context 'when status completed and file exists' do
+      before do
+        allow(bulk_action).to receive(:status).and_return('Completed')
+        allow(File).to receive(:exist?).and_return(true)
+      end
+
+      it 'returns true' do
+        expect(helper.show_report_link?(bulk_action, Settings.checksum_report_job.csv_filename)).to be_truthy
```
Comment (ndushay) [inline spec/helpers/bulk_action_helper_spec.rb]: Things I tried:

```
expect(helper).to show_report_link(bulk_action, Settings.checksum_report_job.csv_filename) # undefined method `show_report_link'
expect(helper).to show_report_link?(bulk_action, Settings.checksum_report_job.csv_filename) # undefined method `matches?' for true:TrueClass
expect(helper).to be_show_report_link(bulk_action, Settings.checksum_report_job.csv_filename) # works but ... yecccch.
```

**Comment by ndushay** [comment]
I would really like to get this deployed today.  The disconnect between where preservation_robots-stage puts stuff and where preservation_catalog-stage looks for it means testing this particular thing in Argo stage wasn't optimal.

File: spec/helpers/bulk_action_helper_spec.rb [hunk 14]
```diff
@@ -32,4 +32,40 @@
...
+    end
+
+    context 'when status completed but file does not exist' do
+      before do
+        allow(bulk_action).to receive(:status).and_return('Completed')
+        allow(File).to receive(:exist?).and_return(false)
+      end
+
+      it 'returns false' do
+        expect(helper.show_report_link?(bulk_action, Settings.checksum_report_job.csv_filename)).to be_falsey
```
Comment (jcoyne) [inline spec/helpers/bulk_action_helper_spec.rb]: another way to do this is:
```
subject { helper.show_report_link?(bulk_action, Settings.checksum_report_job.csv_filename) }

it { is_expected.to be false }

```

Comment (ndushay) [reply spec/helpers/bulk_action_helper_spec.rb]: does this mean you want me to change it?  Is it worth it?

Comment (peetucket) [reply spec/helpers/bulk_action_helper_spec.rb]: i think the current syntax is easier to read (for me at least)

Comment (jmartin-sul) [reply spec/helpers/bulk_action_helper_spec.rb]: i have a slight preference for the way it is at present (seems clearer, though maybe it's less idiomatic?).  but i don't care that much one way or the other.

**Pull request merged**